virttest/qemu_devices/qdevices.py

```diff
@@ -2302,18 +2316,51 @@ def _set_device_props(self, device, addr):
         """
         if self.is_direct_plug(device):
             super(QPCIEBus, self)._set_device_props(device, addr)
+
+    def get_root_port_by_params(self, filt=None):
+        """
+        Get root ports by given dict
+
+        :filt: dict to get root port
+        :return: list of root ports
+        """
+        out = []
+        for root_port in self.__root_ports.values():
+            if filt:
+                for key, value in six.iteritems(filt):
+                    if key not in root_port.params:
+                        continue
+                    if root_port.get_param(key) != value:
+                        continue
+            out.append(root_port)
+        return out
+
+    def prepare_free_root_port(self):
+        """
+        Get a free root port, add a new root port if not found
+
+        :return: QDevice object
+        """
+        root_ports = self.get_root_port_by_params({'multifunction': 'on'})
+        for root_port in sorted(root_ports, key=lambda x: x.get_param('addr')):
+            port_addr = root_port.get_param('addr')
+            last_addr = '%s.7' % port_addr
+            if not self.__root_ports.get(last_addr):
+                break
         else:
-            self.__root_ports[addr[0]].insert(device)
+            root_port = self.add_root_port(self.__root_port_type)
+        root_port = self.add_root_port(self.__root_port_type, root_port)
+        return root_port
 
     def get_free_root_port(self):
         """
-        Get a free slot from any pcie root port
-        :return: The free root port busid
+        Get free root port. address 0x0 of multifunction port is reserved so,
+        multifunction port not be return
```

Comment:
I don't really understand the multifunction note, can you please elaborate a bit? As far as I understand the code I'd wirte `Get a root port with a free slot.`

Reply:
The original function implements the function of returning the pcie bus ID, it doesn't match function name. so I updated it to return a p root port device.

Reply:
update doc string.